core/logger/zap_disk_logging.go

```diff
@@ -0,0 +1,96 @@
+package logger
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/smartcontractkit/chainlink/core/utils"
+
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"gopkg.in/natefinch/lumberjack.v2"
+)
+
+var diskPollInterval = 1 * time.Minute
+
+type zapDiskPollConfig struct {
+	stop     func()
+	pollChan <-chan time.Time
+}
+
+func newDiskPollConfig(interval time.Duration) zapDiskPollConfig {
+	ticker := time.NewTicker(utils.WithJitter(interval))
+
+	return zapDiskPollConfig{
+		pollChan: ticker.C,
+		stop:     ticker.Stop,
+	}
+}
+
+func newDiskCore(cfg ZapLoggerConfig) (zapcore.Core, error) {
+	availableSpace, err := cfg.diskStats.AvailableSpace(cfg.local.Dir)
+	if err != nil {
+		return nil, errors.Wrap(err, "error getting disk space available for logging")
+	}
+	if availableSpace < cfg.local.RequiredDiskSpace {
+		return nil, fmt.Errorf(
+			"disk space is not enough to log into disk, required disk space: %s, Available disk space: %s",
+			cfg.local.RequiredDiskSpace,
+			availableSpace,
+		)
+	}
+
+	var (
+		encoder = zapcore.NewConsoleEncoder(makeEncoderConfig(cfg.local))
+		sink    = zapcore.AddSync(&lumberjack.Logger{
+			Filename:   logFileURI(cfg.local.Dir),
+			MaxSize:    cfg.local.DiskMaxSizeBeforeRotate,
+			MaxAge:     cfg.local.DiskMaxAgeBeforeDelete,
+			MaxBackups: cfg.local.DiskMaxBackupsBeforeDelete,
+			Compress:   true,
+		})
+		allLogLevels = zap.LevelEnablerFunc(cfg.diskLogLevel.Enabled)
+	)
+
+	return zapcore.NewCore(encoder, sink, allLogLevels), nil
+}
+
+func (l *zapLogger) pollDiskSpace() {
+	defer close(l.closeDiskPollChan)
```

Comment:
Usually it is safer to have the sender do the closing instead of the receiver. And in this case, we can just use close() as the signal, without doing an explicit send - the case below will select the same.